### ISSUE TITLE
[Research] Two-Brain Hyperscanning for Social Cognition with Consumer EEG

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#130]
## Research Topic
EEG hyperscanning — simultaneous recording from two people during social interaction — to measure inter-brain synchrony, empathy, and social cognition using consumer-grade headbands.

## Paper References

1. **"Two-brain microstates: A novel hyperscanning-EEG method for quantifying task-driven inter-brain asymmetry"** — Barilari et al., Journal of Neuroscience Methods, December 2024. Introduces a novel method that applies microstate analysis to concatenated two-brain EEG data. Unlike traditional inter-brain synchrony (IBS) measures, two-brain microstates capture quasi-stable moments of inter-brain synchronous activity while allowing quantification of asymmetric brain states (e.g., observer vs. actor roles). Traditional single-brain microstates were NOT modulated by interaction conditions, but two-brain microstates WERE — validating this as a genuinely social-level neural measure.

2. **"Arbitrary methodological decisions skew inter-brain synchronization estimates in hyperscanning-EEG studies"** — Imaging Neuroscience (MIT Press), 2024. Critical methodological paper: showed that preprocessing choices (epoch length, filtering, connectivity metric) dramatically affect IBS estimates. Short epochs (<1 second) produce inflated IBS. Recommends standardized preprocessing pipelines — directly relevant to building a reliable hyperscanning feature in NeuralDreamWorkshop.

3. **"Embodied hyperscanning for studying social interaction: A scoping review"** — Social Neuroscience, November 2024. Reviews 113 studies combining brain + body measurements during social interaction. EEG is the most used brain imaging modality for hyperscanning due to portability and temporal resolution. Key finding: combining EEG with physiological signals (heart rate, galvanic skin response) improves social cognition detection.

4. **"Developing a Multi-site Hyperscanning Procedure for Investigating Interbrain Synchrony Underlying Remote Social Interaction"** — Springer, 2025. Demonstrates remote hyperscanning — two people in different locations with synchronized EEG recording. Used consumer-grade EEG devices. Proves that hyperscanning does not require physical co-location, opening up telehealth and remote collaboration use cases.

5. **"Hyperscanning: from inter-brain coupling to causality"** — Frontiers in Human Neuroscience, 2024. Reviews methods beyond correlation: Granger causality, transfer entropy, and dynamic causal modeling for hyperscanning. Shows directional information flow (leader/follower dynamics) can be extracted from EEG pairs. Frontal theta synchrony is the strongest marker of cooperative engagement.

## Application to NeuralDreamWorkshop

This adds a "Social Brain" mode where two Muse 2 headbands connect to the same session, enabling:
1. Real-time inter-brain synchrony visualization during conversation/therapy
2. Empathy score based on frontal alpha/theta coupling between brains
3. Leader/follower dynamics in collaborative tasks
4. Couples therapy or parent-child interaction monitoring

### Technical feasibility with Muse 2:
- Two Muse 2 devices can connect via BrainFlow simultaneously (separate board instances)
- Each provides 4-channel 256 Hz EEG — sufficient for phase-locking value (PLV) and coherence
- Frontal channels (AF7/AF8) are the most relevant for social cognition hyperscanning
- Time synchronization: BrainFlow timestamps + NTP correction (sub-50ms precision achievable)

### New files:
- `ml/models/hyperscanning_analyzer.py` — HyperscanningAnalyzer class
- `ml/processing/inter_brain_sync.py` — IBS computation utilities
- `ml/hardware/dual_device_manager.py` — manages two simultaneous Muse connections
- `client/src/pages/social-brain.tsx` — dual-brain visualization dashboard

### Key classes/methods:
```python
class HyperscanningAnalyzer:
    """Computes inter-brain synchrony between two EEG streams."""
    
    def __init__(self):
        self.sync_metrics = ['plv', 'coherence', 'cca', 'granger']
        self.epoch_length_sec = 4.0  # per methodological recommendations
    
    def compute_inter_brain_sync(self, eeg_a: np.ndarray, eeg_b: np.ndarray,
                                  fs: int = 256) -> dict:
        """Compute synchrony between two 4-channel EEG streams.
        Returns: plv_by_band, coherence_by_band, global_sync_score,
                 channel_pair_sync (AF7_a-AF7_b, AF8_a-AF8_b, etc.)"""
    
    def compute_empathy_index(self, sync_results: dict) -> float:
        """Frontal alpha/theta inter-brain coupling → empathy score (0-1).
        Based on: frontal theta PLV (cooperative engagement) + 
                  alpha coherence (shared attention state)."""
    
    def detect_leader_follower(self, eeg_a: np.ndarray, eeg_b: np.ndarray,
                                fs: int = 256) -> dict:
        """Granger causality analysis to determine information flow direction.
        Returns: leader (A/B/mutual), causality_strength, direction_by_band"""

class InterBrainSync:
    """Low-level inter-brain synchrony computation utilities."""
    
    def phase_locking_value_cross(self, signal_a: np.ndarray, signal_b: np.ndarray,
                                   band: tuple, fs: int) -> float:
        """Cross-brain PLV in a frequency band."""
    
    def circular_correlation(self, phase_a: np.ndarray, phase_b: np.ndarray) -> float:
        """Corrected circular correlation (per 2024 methodology paper)."""
```

### API endpoints:
- `POST /hyperscanning/start-session` — start dual-device recording
- `GET /hyperscanning/sync-status` — real-time inter-brain synchrony
- `GET /hyperscanning/empathy-score` — computed empathy index
- `GET /hyperscanning/session-report` — post-session social interaction analysis

### Frontend:
- Split-screen dual-brain wave visualization
- Real-time synchrony heatmap (channel-pair × frequency-band)
- Empathy score gauge with trend over session
- Leader/follower flow indicator

## Expected Accuracy/Impact
- Inter-brain PLV detection: reliable with 4-second epochs (per methodological paper)
- Social engagement vs. non-engagement classification: **70-80%** based on frontal theta synchrony
- Empathy index: correlates with behavioral empathy measures at r=0.3-0.5 (moderate)
- Unique differentiator: very few consumer BCI apps offer hyperscanning — this would be a first-of-kind feature
- Use cases: couples therapy, parent-child bonding assessment, team collaboration optimization

## Complexity Rating: HARD
- Requires managing two simultaneous BrainFlow device connections (dual Bluetooth)
- Time synchronization between devices is non-trivial (clock drift correction needed)
- PLV/coherence computation across brains requires careful epoch alignment
- Frontend dual-visualization is significant UI work
- No training data needed — purely signal processing based (PLV, coherence are mathematical)
- Bluetooth bandwidth: two Muse 2 devices may cause RF interference — needs testing
- Estimated implementation: 2-3 weeks for core + API + basic UI
